Scripts/Runtime/Data/VirtualData/VirtualData.cs

```diff
@@ -0,0 +1,296 @@
+        {
+            return new VDJobResultsDestination<TInstance>(m_Pending.AsWriter());
+        }
+
+        internal VDJobUpdater<TKey, TInstance> CreateVDJobUpdater()
+        {
+            return new VDJobUpdater<TKey, TInstance>(m_Pending.AsWriter(),
+                                                     m_Iteration.AsDeferredJobArray(),
+                                                     m_Lookup);
+        }
+
+        internal VDJobWriter<TInstance> CreateVDJobWriter()
+        {
+            return new VDJobWriter<TInstance>(m_Pending.AsWriter());
+        }
+
+        //*************************************************************************************************************
+        // CONSOLIDATION
+        //*************************************************************************************************************
+
+        JobHandle IVirtualData.ConsolidateForFrame(JobHandle dependsOn)
+        {
+            return ConsolidateForFrame(dependsOn);
+        }
+
+        private JobHandle ConsolidateForFrame(JobHandle dependsOn)
+        {
+            JobHandle exclusiveWriteHandle = m_AccessController.AcquireAsync(AccessType.ExclusiveWrite);
+            ConsolidateLookupJob consolidateLookupJob = new ConsolidateLookupJob(m_Pending,
+                                                                                 m_Iteration,
+                                                                                 m_Lookup);
+            JobHandle consolidateHandle = consolidateLookupJob.Schedule(JobHandle.CombineDependencies(dependsOn, exclusiveWriteHandle));
+
+            m_AccessController.ReleaseAsync(consolidateHandle);
+
+            return consolidateHandle;
+        }
+
+        //*************************************************************************************************************
+        // JOBS
+        //*************************************************************************************************************
+
+        [BurstCompile]
+        private struct ConsolidateLookupJob : IJob
+        {
+            private UnsafeTypedStream<TInstance> m_Pending;
+            private DeferredNativeArray<TInstance> m_Iteration;
+            private UnsafeHashMap<TKey, TInstance> m_Lookup;
+
+            public ConsolidateLookupJob(UnsafeTypedStream<TInstance> pending,
+                                        DeferredNativeArray<TInstance> iteration,
+                                        UnsafeHashMap<TKey, TInstance> lookup)
+            {
+                m_Pending = pending;
+                m_Iteration = iteration;
+                m_Lookup = lookup;
+            }
+
+            public void Execute()
+            {
+                m_Lookup.Clear();
+                m_Iteration.Clear();
+
+                NativeArray<TInstance> iterationArray = m_Iteration.DeferredCreate(m_Pending.Count());
+                m_Pending.CopyTo(ref iterationArray);
+                m_Pending.Clear();
```

Comment:
Sure let's chat tomorrow. 

Here's what's happening:

- The previous frame's data is out of date so we clear both our Lookup and our Iteration collections. 
- Now we allocate the memory needed for the total amount of elements by looking at the count in Pending. 
    - Pending represents any new elements AND any elements from last frame that were `Continue`ed.
- With the array sized, we use the fast `CopyTo` which blits the blocks of memory to the array instead of going one by one.
- We then clear Pending because we'll want to write to it again this frame.
- We then need to populate the Lookup. But the lookup doesn't have a fast blit method and we can't bypass it either because it stores things in blocks. 
    - Array iteration is super fast and we already have our array so we loop through it and use the fastest add function for adding to the lookup.
- We now have a Lookup and an Iteration that contain all the same elements and an empty primed Pending collection. We are ready for any jobs this frame that need to read or update or add new elements or continue existing ones to the next frame.

Reply:
hmmm, probably not worth testing this micro optimization but since you have to perform a for iteration anyway I wonder if Blit Copy is actually faster than assigning one element at a time.

Also for another day, `UnsafeHashMap` doesn't have a bulk add method but I bet one could be implemented that mem copied the entire value array and then did a mem copy with a stride to extract the keys.

Reply:
ah ok, makes sense.

Yea hold off on HashMap until some other day. Maybe make an issue if you really want to get into it some day.